runtime/runtime/src/receipt_manager.rs

```diff
@@ -145,29 +148,63 @@ impl ReceiptManager {
             actions: vec![],
             is_promise_yield: true,
         };
-        let new_receipt_index = self.action_receipts.len() as ReceiptIndex;
+        let new_receipt_index = self.action_receipts.len();
         self.action_receipts.push(new_receipt);
-        self.yielded_data_ids.insert((input_data_id, receiver_id));
-        Ok(new_receipt_index)
+        self.promise_yield_receipt_index.insert(input_data_id, new_receipt_index);
+        Ok(new_receipt_index as ReceiptIndex)
     }
 
-    /// Submits a data receipt.
+    /// Creates a PromiseResume receipt.
     ///
     /// Should only be used to resolve dependencies created by `create_yielded_action_receipt`.
     ///
     /// # Arguments
     ///
-    /// * `data_id` - id of the DataReceipt being submitted
-    /// * `data` - contents of the DataReceipt
-    pub(super) fn submit_promise_resume_data(
+    /// * `data_id` - id of the PromiseResume receipt being submitted
+    /// * `data` - contents of the PromiseResume receipt
+    pub(super) fn create_promise_resume_receipt(
         &mut self,
         data_id: CryptoHash,
         data: Vec<u8>,
     ) -> Result<(), VMLogicError> {
-        self.data_receipts.push(DataReceiptMetadata { data_id, data: Some(data) });
+        self.data_receipts.push(DataReceiptMetadata {
+            data_id,
+            data: Some(data),
+            is_promise_resume: true,
+        });
         Ok(())
     }
 
+    /// Resolves a PromiseYield input dependency previously created under given `data_id`,
+    /// if it exists.
+    ///
+    /// # Arguments
+    ///
+    /// * `data_id` - id of the Data receipt being submitted
+    /// * `data` - contents of the Data receipt
+    pub(super) fn checked_resolve_promise_yield(
+        &mut self,
+        data_id: CryptoHash,
+        data: Vec<u8>,
+    ) -> Result<bool, VMLogicError> {
+        Ok(if let Some(receipt_index) = self.promise_yield_receipt_index.remove(&data_id) {
+            // Convert existing PromiseYield to a standard Action receipt
+            let receipt = &mut self.action_receipts[receipt_index];
+            assert!(receipt.is_promise_yield, "receipt should be promise yield");
+            receipt.is_promise_yield = false;
+
+            // Create Data receipt delivering the payload
+            self.data_receipts.push(DataReceiptMetadata {
```

Comment:
I wonder if we could possibly get away somehow without involving data receipts at all here... Probably not since we pass information around as promise inputs, and it would require some special casing to enable…?

Reply:
I guess we could directly process the data receipt (knowing the action receipt will arrive later) and set the ReceivedData key directly in the trie ([see link](https://github.com/near/nearcore/blob/00d51bde97378995d2adc058fa30fe299e18dc38/runtime/runtime/src/lib.rs#L913-L918)).